### ISSUE TITLE
HonoClientImpl may reach state where getOrCreateSender() fails each time

### ISSUE THREAD

**Issue by calohmn** [eclipse/hono#159]
We noticed that at one point our Hono REST adapter instance began answering each telemetry request with a 503 response.
A look in the logs revealed the following:

```
07:19:14.755 [vert.x-eventloop-thread-0] DEBUG o.e.hono.client.impl.HonoClientImpl - creating new message sender for telemetry/DEFAULT_TENANT
07:19:14.759 [vert.x-eventloop-thread-0] DEBUG o.e.hono.client.impl.AbstractSender - sender open [Target{address='telemetry/DEFAULT_TENANT', durable=NONE, expiryPolicy=SESSION_END, timeout=0, dynamic=false, dynamicNodeProperties=null, capabilities=null}]
07:19:14.759 [vert.x-eventloop-thread-0] DEBUG o.e.hono.client.impl.HonoClientImpl - successfully created new message sender for telemetry/DEFAULT_TENANT

07:19:24.745 [vert.x-eventloop-thread-0] DEBUG o.e.hono.client.impl.HonoClientImpl - reusing existing message sender [target: telemetry/DEFAULT_TENANT, credit: 250]

07:19:24.748 [vert.x-eventloop-thread-0] DEBUG o.e.hono.client.impl.AbstractSender - sender [telemetry/DEFAULT_TENANT] closed: Error{condition=amqp:precondition-failed, description='device does not exist', info=null}

07:19:34.844 [vert.x-eventloop-thread-0] DEBUG o.e.hono.client.impl.HonoClientImpl - creating new message sender for telemetry/DEFAULT_TENANT

07:20:50.584 [vert.x-eventloop-thread-0] DEBUG o.e.h.a.r.VertxBasedRestProtocolAdapter - registering data for device: {"creater":"self-registered"}
07:20:50.584 [vert.x-eventloop-thread-0] DEBUG o.e.hono.client.impl.HonoClientImpl - reusing existing registration client for [DEFAULT_TENANT]
07:20:50.588 [vert.x-eventloop-thread-0] DEBUG o.e.h.c.impl.RegistrationClientImpl - received response [correlation ID: reg-client-15, status: 201]
07:20:56.150 [vert.x-eventloop-thread-0] DEBUG o.e.hono.client.impl.HonoClientImpl - already trying to create a message sender for telemetry/DEFAULT_TENANT
07:21:06.232 [vert.x-eventloop-thread-0] DEBUG o.e.hono.client.impl.HonoClientImpl - already trying to create a message sender for telemetry/DEFAULT_TENANT
07:21:16.361 [vert.x-eventloop-thread-0] DEBUG o.e.hono.client.impl.HonoClientImpl - already trying to create a message sender for telemetry/DEFAULT_TENANT
07:21:26.358 [vert.x-eventloop-thread-0] DEBUG o.e.hono.client.impl.HonoClientImpl - already trying to create a message sender for telemetry/DEFAULT_TENANT
07:21:36.439 [vert.x-eventloop-thread-0] DEBUG o.e.hono.client.impl.HonoClientImpl - already trying to create a message sender for telemetry/DEFAULT_TENANT
07:21:46.547 [vert.x-eventloop-thread-0] DEBUG o.e.hono.client.impl.HonoClientImpl - already trying to create a message sender for telemetry/DEFAULT_TENANT
07:21:56.671 [vert.x-eventloop-thread-0] DEBUG o.e.hono.client.impl.HonoClientImpl - already trying to create a message sender for telemetry/DEFAULT_TENANT
...
```
The last message is repeated with each new telemetry request.

It looks like the last invocation of _HonoClientImpl.getOrCreateSender()_ where a new Sender was about to be created ("creating new message sender for telemetry/DEFAULT_TENANT") did not finish.
That means each following invocation fails ("already trying to create a message sender"). 

Maybe there should be a timeout associated with the sender creation attempt?

**Comment by sophokles73** [comment]
What version are you using?

**Comment by calohmn** [comment]
Release 0.5-M5
(last contained commit is "Bump version to 0.5-M6-SNAPSHOT" from 2017-04-07)

**Comment by sophokles73** [comment]
> It looks like the last invocation of HonoClientImpl.getOrCreateSender() where a new Sender was about to be created ("creating new message sender for telemetry/DEFAULT_TENANT") did not finish.

Does this mean that you were expecting the attempt to create the sender to fail (fast), i.e. was your Hono server down at that time?

**Comment by calohmn** [comment]
No, the Hono server was not down. But it looks like the downstream connection was lost.

Excerpt from Hono server log:
```
07:19:34.845 [vert.x-eventloop-thread-1] DEBUG org.eclipse.hono.server.HonoServer - client [Hono REST Adapter-25e63966-1770-43bf-b79c-a13b7163aa61] wants to open a link for sending messages [address: Target{address='telemetry/DEFAULT_TENANT', durable=NONE, expiryPolicy=SESSION_END, timeout=0, dynamic=false, dynamicNodeProperties=null, capabilities=null}]
07:19:34.845 [vert.x-eventloop-thread-0] DEBUG o.e.h.a.i.BaseAuthorizationService - subject [rest-adapter] is allowed to WRITE on resource [telemetry/DEFAULT_TENANT]
07:19:34.854 [vert.x-eventloop-thread-1] WARN o.e.h.t.i.ForwardingTelemetryDownstreamAdapter - lost connection to downstream container [Hono.Example.Router], closing upstream receivers ...
07:19:34.859 [vert.x-eventloop-thread-1] DEBUG o.e.h.t.impl.TelemetryEndpoint - closing receiver for client [040ef25f-e7eb-4f2d-a025-3d6d051e1cd8]

07:19:35.157 [vert.x-eventloop-thread-1] INFO o.e.h.t.i.ForwardingTelemetryDownstreamAdapter - attempting to re-connect to downstream container
..
07:19:35.157 [vert.x-eventloop-thread-1] INFO o.e.h.c.ConnectionFactoryImpl - connecting to AMQP 1.0 container [amqps://qdrouter:5673]
07:19:35.222 [vert.x-eventloop-thread-1] INFO o.e.h.c.ConnectionFactoryImpl - connected to AMQP 1.0 container [amqps://qdrouter:5673], opening connection ...
07:19:35.224 [vert.x-eventloop-thread-1] INFO o.e.h.c.ConnectionFactoryImpl - connection to container [Hono.Example.Router] at [amqps://qdrouter:5673] open
```

**Comment by sophokles73** [comment]
should be fixed in 80b8a9cffae42bbde5bf6e443ff63972a96eddd2

@calohmn please re-open if the problem occurs again.
